// react-color-wheel/lib/HueWheelDefs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _chromaJs = require('chroma-js');

var _chromaJs2 = _interopRequireDefault(_chromaJs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var HueWheelDefs = function (_React$PureComponent) {
  _inherits(HueWheelDefs, _React$PureComponent);

  function HueWheelDefs() {
    _classCallCheck(this, HueWheelDefs);

    return _possibleConstructorReturn(this, (HueWheelDefs.__proto__ || Object.getPrototypeOf(HueWheelDefs)).apply(this, arguments));
  }

  _createClass(HueWheelDefs, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          n = _props.n,
          componentId = _props.componentId;

      var items = [];
      for (var i = 0; i < n; i++) {
        var ref = i * 2 * Math.PI / n;
        var lo = ref;
        var loDeg = lo * 180 / Math.PI;
        var hi = ref + 2 * Math.PI / n;
        var hiDeg = hi * 180 / Math.PI;

        var r1 = ref + Math.PI / n;

        var x10 = Math.sin(r1);var xa = Math.abs(x10);
        var y10 = Math.cos(r1);var ya = Math.abs(y10);
        var x20 = -Math.sin(r1);
        var y20 = -Math.cos(r1);
        var mult = 1 / Math.max(xa, ya);

        var x11 = (mult * x10 + 1) / 2;
        var y11 = (mult * y10 + 1) / 2;
        var x21 = (mult * x20 + 1) / 2;
        var y21 = (mult * y20 + 1) / 2;

        items.push(_react2.default.createElement(
          'linearGradient',
          {
            key: i,
            id: 'g' + i + componentId,
            gradientUnits: 'objectBoundingBox',
            x1: x11,
            y1: y11,
            x2: x21,
            y2: y21
          },
          _react2.default.createElement('stop', { offset: '0%', stopColor: '' + _chromaJs2.default.hsv(loDeg, 1, 1).hex() }),
          _react2.default.createElement('stop', { offset: '100%', stopColor: '' + _chromaJs2.default.hsv(hiDeg, 1, 1).hex() })
        ));
      }

      return _react2.default.createElement(
        'defs',
        null,
        items
      );
    }
  }]);

  return HueWheelDefs;
}(_react2.default.PureComponent);

exports.default = HueWheelDefs;
